### PR TITLE
Introduce zod for kt forms and fields

### PR DESCRIPTION
**Pull request by carsoli**
TODO: Change logs

### CODE REVIEW

**Comment by carsoli** [comment]
revert https://github.com/3YOURMIND/kotti/pull/576/commits/908f414e8cb19792ed377f3e7454527b9fe01e39?authenticity_token=C8DbDpSxQhYUhA0r%2F9Q2evSQqRHZ%2FrAIKI0lZtjRq7ZZvv%2Ff9E57hIlTreyzuu9bQOQIyYzmJ%2BjWwPmACl0yrw%3D%3D&file-filters%5B%5D=.ts


- we want `placeholder` on the supports feature 
-  we then would bring back the unknown / otherwise we would need to suppress the type on `useSupports`
- for the fields that don't extend it, we would need to either make `makeProps` accept unknown or maybe support an argument on makeProps that doesn't check for certain strings at all (`placeholder`) in our case (TBD)


Contrast the solution with what we do for `decoration` supports (`prefix`, `suffix`, `rightIcon`, `leftIcon`) because they now default to string nullable, but we don't accept them on toggle field for instance 🤷🏼

File: packages/kotti-ui/source/kotti-field/tests.ts [hunk 1]
```diff
@@ -346,105 +281,6 @@ describe('useField', () => {
...
-									type: 'warning',
-									text: 'This is a warning',
-								}),
-							},
-							values: { testKey: 'something' },
-						}),
-					},
-				})
-
-				forceVueToEvaluateComputedProperty(wrapper.vm.field.validation)
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/tests.ts:373]: Unused now (eslint error)

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 2]
```diff
@@ -38,196 +83,199 @@ export namespace KottiField {
...
+	 * 	placeholder: ['placeholder']
+	 * }
+	 *
+	 * The schema is only concerned with props that would have the same type,
+	 * if extended (e.g. decoration props).
+	 * Therefore, doesn't include `placeholder` since it is extended
+	 * differently per field.
+	 *
+	 */
+
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/types.ts:121]:
```suggestion
```

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 3]
```diff
@@ -38,196 +83,199 @@ export namespace KottiField {
...
+	 * 	decoration: ['leftIcon', 'rightIcon', 'prefix', 'suffix']
+	 * 	tabIndex: ['tabIndex']
+	 * 	placeholder: ['placeholder']
+	 * }
+	 *
+	 * The schema is only concerned with props that would have the same type,
+	 * if extended (e.g. decoration props).
+	 * Therefore, doesn't include `placeholder` since it is extended
+	 * differently per field.
+	 *
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/types.ts:119]:
```suggestion
```

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 4]
```diff
@@ -38,196 +83,199 @@ export namespace KottiField {
...
 		/**
 		 * Defines the size of the field which influences child styles
 		 * to make fields e.g. appear more compact
 		 */
-		size: Size | null
-	}
-
+		size: z.nativeEnum(Size).default(Size.MEDIUM),
+	})
+	export type InhertiableProps = z.input<typeof inheritablePropsSchema>
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/types.ts:166]:
```suggestion

	export type InhertiableProps = z.input<typeof inheritablePropsSchema>
```

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 5]
```diff
@@ -38,196 +83,199 @@ export namespace KottiField {
...
 	/**
-	 * When adding a new prop, please make sure that no KtFormField
-	 * already uses a prop with the same name, to avoid conflicts
+	 * Warning: All keys of type `never` are replaced with `any`
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/types.ts:168]:
```suggestion
	 * Warning: All values of type `never` are replaced with `any`
```

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 6]
```diff
@@ -38,196 +83,199 @@ export namespace KottiField {
...
-		LARGE = 'large',
-		MEDIUM = 'medium',
-		SMALL = 'small',
-	}
+		value: z.never(),
+	})
+	export type Props = z.input<typeof propsSchema>
+	/**
+	 * Warning: All keys of type `never` are replaced with `any`
+	 */
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/types.ts:244]:
```suggestion

	/**
	 * Warning: All values of type `never` are replaced with `any`
	 */
```

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 7]
```diff
@@ -38,196 +83,199 @@ export namespace KottiField {
...
+	export type PropsInternal = ReplaceRecordType<z.output<typeof propsSchema>>
 
 	/**
 	 * Object that explicitly specifies which of the KtField.Props are
 	 * supported, since it may not always make sense to have some of
 	 * the props on certain fields.
 	 */
-	export type Supports = {
+	export const supportsSchema = z.object({
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/types.ts:252]: I think there’s no need to use `zod` here

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 8]
```diff
@@ -38,196 +83,199 @@ export namespace KottiField {
...
-			| KottiField.Validation.Warning
-
-		// eslint-disable-next-line @typescript-eslint/no-explicit-any
-		export type Function = (value: any) => KottiField.Validation.Result
-	}
+		tabIndex: z.boolean(),
+	})
+	export type Supports = z.output<typeof supportsSchema>
+
+	export const translationsSchema = z.object({
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field/types.ts:276]: I think there’s no need to use `zod` here

File: packages/kotti-ui/source/kotti-field-date/types.ts [hunk 9]
```diff
@@ -1,45 +1,118 @@
...
+
+		/**
+		 * [ ISO8601, ISO8601 ]
+		 */
+		export type Value = z.output<typeof valueSchema>
+
+		export const propsSchema = Shared.propsSchema.extend({
+			placeholder: z
+				.tuple([z.string().nullable(), z.string().nullable()])
+				.default([null, null]),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field-date/types.ts:75]:
```suggestion
				.default(() => [null, null]),
```

File: packages/kotti-ui/source/kotti-field-date/types.ts [hunk 10]
```diff
@@ -1,45 +1,118 @@
...
+			shortcuts: z.array(Range.dateShortcutSchema).default(() => []),
+			value: Range.valueSchema.default([null, null]),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-field-date/types.ts:77]:
```suggestion
			value: Range.valueSchema.default(() => [null, null]),
```

**Review by FlorianWendelborn — CHANGES_REQUESTED** [review]

File: packages/kotti-ui/source/kotti-field-toggle/types.ts [hunk 11]
```diff
@@ -1,26 +1,64 @@
...
 
-	export type Props = KottiField.Props<Value, never> & {
-		isInline: boolean
-		options: Entry[]
-		type: 'checkbox' | 'switch'
-	}
+	export const entrySchema = z.object({
+		isDisabled: z.boolean().optional(),
+		// equivalent to NonNullable<valueSchema>.keySchema
+		key: valueSchema.unwrap().keySchema,
```
[the remark this answers is not in the document]
Comment (carsoli) [reply packages/kotti-ui/source/kotti-field-toggle/types.ts:42]: why?

File: packages/kotti-ui/source/kotti-field-date/types.ts (hunk 9 above)
Comment (carsoli) [reply packages/kotti-ui/source/kotti-field-date/types.ts:75]: Unlike arrays, tuples don't accept it

Zod errors if you do that (try it out locally)

File: packages/kotti-ui/source/kotti-field-date/types.ts (hunk 10 above)
Comment (carsoli) [reply packages/kotti-ui/source/kotti-field-date/types.ts:77]: See https://github.com/3YOURMIND/kotti/pull/576#discussion_r785907404

File: packages/kotti-ui/source/kotti-field-toggle/types.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (carsoli) [reply packages/kotti-ui/source/kotti-field-toggle/types.ts:40]: wanted clarity on it being the key of the record, and thought it'd somehow provide extra validation at runtime

don't mind if the 2nd point doesn't hold (RT validation) and don't mind to state that the key must match the record 

back to `z.string()`

File: packages/kotti-ui/source/kotti-field-date/types.ts [hunk 12]
```diff
@@ -1,45 +1,118 @@
...
+			]),
+		})
+		export const dateTimeShortcutSchema = Shared.dateShortcutSchema.extend({
+			value: z.tuple([
+				z.string().regex(DATE_TIME_FORMAT_REGEX),
+				z.string().regex(DATE_TIME_FORMAT_REGEX),
+			]),
+		})
+
+		export const valueSchema = z.tuple([
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field-date/types.ts:62]: needs the dayjs check for format

File: packages/kotti-ui/source/kotti-field-date/types.ts [hunk 13]
```diff
@@ -1,45 +1,118 @@
...
+	export const dateShortcutSchema = z.object({
+		keepOpen: z.boolean().optional(),
+		label: z.string(),
+		value: z.string().regex(DATE_FORMAT_REGEX),
+	})
+	export const dateTimeShortcutSchema = Shared.dateShortcutSchema.extend({
+		value: z.string().regex(DATE_TIME_FORMAT_REGEX),
+	})
 
+	export const valueSchema = z.string().nullable()
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field-date/types.ts:18]: DATE_FORMAT_REGEX

File: packages/kotti-ui/source/kotti-field-number/utilities.ts [hunk 14]
```diff
@@ -22,8 +22,8 @@ export const isInRange = ({
 	minimum,
 	value,
 }: {
-	maximum: KottiFieldNumber.Props['maximum']
-	minimum: KottiFieldNumber.Props['minimum']
+	maximum: KottiFieldNumber.PropsInternal['maximum']
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field-number/utilities.ts:25]: no consistency in those utils

sometimes, we make util generically support `number | null`

in this case, we rely on PropsInternal 

unify as generic `number | null`

File: packages/kotti-ui/source/kotti-field-select/KtFieldMultiSelect.vue [hunk 15]
```diff
@@ -86,28 +86,23 @@ import { Select as ElSelect, Option as ElOption } from 'element-ui'
...
 type Entry = KottiFieldMultiSelect.Props['options'][0]
 
-export default defineComponent({
+export default defineComponent<KottiFieldMultiSelect.PropsInternal>({
 	name: 'KtFieldMultiSelect',
 	components: { ElOption, ElSelect, KtField, ActionIcon },
-	props: {
-		...KOTTI_FIELD_PROPS,
-		...KOTTI_FIELD_MULTI_SELECT_PROPS,
-		value: { default: null, type: Array },
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field-select/KtFieldMultiSelect.vue:107]: make sure default is fixed

**Review by FlorianWendelborn — CHANGES_REQUESTED** [review]
Actionbar needs zod label

File: packages/kotti-ui/source/kotti-field-select/types.ts [hunk 16]
```diff
@@ -45,24 +52,57 @@ export namespace Shared {
...
+	export type Value = z.output<typeof valueSchema>
 
-	export type Value = Shared.Entry['value']
+	export const propsSchema = KottiField.propsSchema
+		.merge(Shared.propsSchema)
+		.extend({
+			value: valueSchema.default(null),
+		})
+
+	export type PropsInternal = z.output<typeof propsSchema>
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field-select/types.ts:83]: unify the order of PropsInternal with Props everywhere

not a blocker

File: packages/kotti-ui/source/kotti-field-toggle/types.ts [hunk 17]
```diff
@@ -1,26 +1,61 @@
...
 
 export namespace KottiFieldToggle {
-	export type Props = KottiField.Props<Value, never> & {
-		type: 'checkbox' | 'switch'
+	export namespace Shared {
+		export enum Type {
+			SWITCH = 'switch',
+			CHECKBOX = 'checkbox',
+		}
+		export const propsSchema = z
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field-toggle/types.ts:12]: extend the `KottiField.propsSchema`

File: packages/kotti-ui/source/kotti-field/constants.ts [hunk 18]
```diff
@@ -1,45 +1,4 @@
...
-
-import { KottiField } from './types'
-
-export const KOTTI_FIELD_INHERITABLE_PROPS = {
-	hideClear: { default: false, type: Boolean },
-	hideValidation: { default: false, type: Boolean },
-	isDisabled: { default: false, type: Boolean },
-	isLoading: { default: false, type: Boolean },
-	size: {
-		default: null,
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field/constants.ts:11]: medium

Comment (carsoli) [reply packages/kotti-ui/source/kotti-field/constants.ts:11]: bring back the null to distinguish the size passed down from context  as default null

File: packages/kotti-ui/source/kotti-field/tests.ts [hunk 19]
```diff
@@ -5,21 +5,18 @@ import { shallowMount } from '@vue/test-utils'
...
-import { KOTTI_FIELD_PROPS, FORM_KEY_NONE } from './constants'
+import { FORM_KEY_NONE } from './constants'
 import { KtFieldErrors } from './errors'
 import { useField } from './hooks'
 import { KottiField } from './types'
 
 const TestComponent = defineComponent({
 	name: 'TestComponent',
-	props: KOTTI_FIELD_PROPS,
-	setup: (props: KottiField.Props<string | null, string | null>, { emit }) => {
```
[the remark this answers is not in the document]
Comment (carsoli) [reply packages/kotti-ui/source/kotti-field/tests.ts:22]: & { value: whatever }

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 20]
```diff
@@ -38,135 +83,166 @@ export namespace KottiField {
...
+
 	/**
 	 * Includes, but is not limited to, properties that are not consumed by the KtForm itself,
 	 * but rather just passed-down to all fields inside the form to make common
 	 * tasks like disabling the entire form or hiding validation easier.
 	 *
 	 * These values can be explicitly overriden in the field usage
 	 */
-	export type InhertiableProps = {
+	export const inheritablePropsSchema = z.object({
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field/types.ts:138]: TYPO: fix all mentions of `inhertiable` to be `Inheritable`

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 21]
```diff
@@ -38,135 +83,166 @@ export namespace KottiField {
...
 		/**
 		 * Shows a Yoco icon on the right side of the field
 		 */
-		rightIcon: Yoco.Icon | null
+		rightIcon: z.never(),
 
-		isOptional: boolean
-		label: string | null
+		isOptional: z.boolean().default(false),
+		label: z.string().nullable().default(null),
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field/types.ts:214]:
```suggestion

		label: z.string().nullable().default(null),
```

File: packages/kotti-ui/source/kotti-field/types.ts [hunk 22]
```diff
@@ -38,135 +83,166 @@ export namespace KottiField {
...
 
 		/**
 		 * Show some string before the field
 		 */
-		prefix: string | null
+		prefix: z.never(),
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field/types.ts:219]: maybe leave a comment for 
`prefix`, `suffix`, `leftIcon`, `rightIcon`


to clarify that it's `never` and specified by the potentiallySupportedPropsSchema IF the component supports them

File: packages/kotti-ui/source/kotti-form-controller-list/tests.ts [hunk 23]
```diff
@@ -57,7 +57,7 @@ const getField = (
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
 	wrapper: Wrapper<any>,
 	index: number,
-): KottiField.Hook.Returns<string | null, string | null> =>
+): KottiField.Hook.Returns<string | null> =>
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-form-controller-list/tests.ts:60]: supports value `string | null` even tho it's never on KottiField

extend the props passed in tests

File: packages/kotti-ui/source/kotti-form-controller-object/tests.ts [hunk 24]
```diff
@@ -57,7 +57,7 @@ const getField = (
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
 	wrapper: Wrapper<any>,
 	index: number,
-): KottiField.Hook.Returns<string | null, string | null> =>
+): KottiField.Hook.Returns<string | null> =>
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-form-controller-object/tests.ts:60]: same comment about the value

File: packages/kotti-ui/source/kotti-form/tests.ts [hunk 25]
```diff
@@ -38,17 +38,11 @@ const TestField = defineComponent({
 	template: `<KtField :field="field" :getEmptyValue="() => null">FIELD</KtField>`,
 })
 
-const TestFieldObject = defineComponent({
+const TestFieldObject = defineComponent<KottiField.PropsInternal>({
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-form/tests.ts:41]: what even is a TestFieldObject??
is this a controller object?

if not, then why is the type `Record<string, unknown>` in the old code's DATA_TYPE

File: packages/kotti-ui/source/kotti-form/tests.ts [hunk 26]
```diff
@@ -38,17 +38,11 @@ const TestField = defineComponent({
...
-			Record<string, unknown> | string | null,
-			string | null
-		>,
-		{ emit },
-	) => {
+	props: makeProps(KottiField.propsSchema),
+	setup: (props, { emit }) => {
 		useI18nProvide(ref('en-US'), ref({}), ref({}))
 
 		return {
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-form/tests.ts:48]: why are we testing the value to be `string | null | Record<string, unknown>`

File: packages/kotti-ui/source/make-props.ts [hunk 27]
```diff
@@ -113,10 +122,14 @@ const walkSchemaTypes = <SCHEMA extends z.ZodTypeAny>(
 		case z.ZodFirstPartyTypeKind.ZodBoolean:
 		case z.ZodFirstPartyTypeKind.ZodDate:
 		case z.ZodFirstPartyTypeKind.ZodFunction:
+		case z.ZodFirstPartyTypeKind.ZodLiteral:
```
Comment (carsoli) [inline packages/kotti-ui/source/make-props.ts:125]: try removing it - might be unneeded if all direct Literal types were now changed to NativeEnum

File: packages/kotti-ui/source/make-props.ts [hunk 28]
```diff
@@ -113,10 +122,14 @@ const walkSchemaTypes = <SCHEMA extends z.ZodTypeAny>(
...
 		case z.ZodFirstPartyTypeKind.ZodNativeEnum:
+		case z.ZodFirstPartyTypeKind.ZodNever: // special case for abstract props
 		case z.ZodFirstPartyTypeKind.ZodNumber:
 		case z.ZodFirstPartyTypeKind.ZodObject:
-		case z.ZodFirstPartyTypeKind.ZodString: {
+		case z.ZodFirstPartyTypeKind.ZodRecord:
```
Comment (carsoli) [inline packages/kotti-ui/source/make-props.ts:130]: test

File: packages/kotti-ui/source/make-props.ts [hunk 29]
```diff
@@ -113,10 +122,14 @@ const walkSchemaTypes = <SCHEMA extends z.ZodTypeAny>(
...
+		case z.ZodFirstPartyTypeKind.ZodString:
+		case z.ZodFirstPartyTypeKind.ZodTuple: {
```
Comment (carsoli) [inline packages/kotti-ui/source/make-props.ts:132]: test

**Comment by carsoli** [comment]
TODO: add `zod` labels to docs

File: packages/kotti-ui/source/kotti-field-select/types.ts [hunk 30]
```diff
@@ -1,41 +1,48 @@
...
+		.merge(
+			KottiField.potentiallySupportedPropsSchema.pick({
+				hideClear: true,
+				leftIcon: true,
+				prefix: true,
+				rightIcon: true,
+				suffix: true,
+			}),
+		)
+	export type Props = z.input<typeof propsSchema>
```
Comment (carsoli) [inline packages/kotti-ui/source/kotti-field-select/types.ts:44]:
```suggestion

	export type Props = z.input<typeof propsSchema>
```

**Review by FlorianWendelborn — APPROVED** [review]

**Pull request merged**